jbpm-task-assigning/jbpm-task-assigning-core/src/test/java/org/jbpm/task/assigning/model/solver/realtime/BaseProblemFactChangeTest.java

```diff
@@ -0,0 +1,149 @@
+
+        private TaskAssigningSolution solutionAfterChange;
+
+        private C change;
+
+        public ProgrammedProblemFactChange() {
+        }
+
+        public ProgrammedProblemFactChange(C change) {
+            this.change = change;
+        }
+
+        public TaskAssigningSolution getSolutionAfterChange() {
+            return solutionAfterChange;
+        }
+
+        public void setSolutionAfterChange(TaskAssigningSolution solutionAfterChange) {
+            this.solutionAfterChange = solutionAfterChange;
+        }
+
+        public C getChange() {
+            return change;
+        }
+
+        public void setChange(C change) {
+            this.change = change;
+        }
+    }
+
+    protected TaskAssigningSolution executeSequentialChanges(TaskAssigningSolution solution, List<? extends ProgrammedProblemFactChange> changes) throws Exception {
+        Solver<TaskAssigningSolution> solver = createDaemonSolver();
+
+        //store the first solution that was produced by the solver for knowing how things looked like at the very
+        //beginning before any change was produced.
+        final TaskAssigningSolution[] initialSolution = {null};
+        final boolean[] changesInProgress = {false};
+
+        final Semaphore programNextChange = new Semaphore(0);
+        final Semaphore allChangesWereProduced = new Semaphore(0);
+        final Semaphore executorThreadStarted = new Semaphore(0);
+
+        //prepare the list of changes to program
+        List<ProgrammedProblemFactChange> programmedChanges = new ArrayList<>(changes);
+        List<ProgrammedProblemFactChange> scheduledChanges = new ArrayList<>();
+
+        int totalProgrammedChanges = programmedChanges.size();
+        int[] pendingChanges = {programmedChanges.size()};
+
+        solver.addEventListener(event -> {
+            if (initialSolution[0] == null) {
+                //store the first produced solution for knowing how things looked like at the very beginning.
+                initialSolution[0] = event.getNewBestSolution();
+                //let the problem fact changes start being produced.
+                programNextChange.release();
+            } else if (changesInProgress[0] && event.isEveryProblemFactChangeProcessed()) {
+                changesInProgress[0] = false;
+                ProgrammedProblemFactChange programmedChange = scheduledChanges.get(scheduledChanges.size() - 1);
+                programmedChange.setSolutionAfterChange(event.getNewBestSolution());
+
+                if (pendingChanges[0] > 0) {
+                    //let the Programmed changes producer produce next change
+                    programNextChange.release();
+                } else {
+                    solver.terminateEarly();
+                    allChangesWereProduced.release();
+                }
+            }
+        });
+
+        //Programmed changes producer Thread.
+        ExecutorService executor = Executors.newSingleThreadExecutor();
```

Comment:
I think that at the end of the day when we reach let's say the last line of the test, the only "Runnable" command that the Executor is running has finished, since the Test can only finish when all the changes were produced.
So the only active thread that the executor had, was already finished his work. So at this point when the test finish and the JVM finishes the executor will have a graceful die.

But your suggestion is good. +1